pytorch/pytorch_ddp_1machine.py

```diff
@@ -0,0 +1,216 @@
+from torch.nn.parallel import DistributedDataParallel as DDP
+import torch.optim as optim
+import torch.utils.data
+from torchvision import datasets
+from torchvision import transforms
+
+
+# Arbitrary GPUs can be selected
+DEVICE_IDS = [0, 1]
+BATCHSIZE = 128
+CLASSES = 10
+DIR = os.getcwd()
+EPOCHS = 10
+N_TRAIN_EXAMPLES = BATCHSIZE * 30
+N_VALID_EXAMPLES = BATCHSIZE * 10
+N_TRIALS = 20
+
+
+def define_model(trial):
+    # We optimize the number of layers, hidden units and dropout ratio in each layer.
+    n_layers = trial.suggest_int("n_layers", 1, 3)
+    layers = nn.ModuleList()
+
+    in_features = 28 * 28
+    for i in range(n_layers):
+        out_features = trial.suggest_int("n_units_l{}".format(i), 4, 128)
+        layers.append(nn.Linear(in_features, out_features))
+        layers.append(nn.ReLU())
+        p = trial.suggest_float("dropout_l{}".format(i), 0.2, 0.5)
+        layers.append(nn.Dropout(p))
+
+        in_features = out_features
+    layers.append(nn.Linear(in_features, CLASSES))
+    layers.append(nn.LogSoftmax(dim=1))
+
+    return nn.Sequential(*layers)
+
+
+def get_mnist():
+    # Load FashionMNIST dataset.
+    train_dataset = datasets.FashionMNIST(DIR, train=True, transform=transforms.ToTensor())
+    train_dataset = torch.utils.data.Subset(train_dataset, indices=range(N_TRAIN_EXAMPLES))
+    train_sampler = torch.utils.data.distributed.DistributedSampler(dataset=train_dataset)
+
+    valid_dataset = datasets.FashionMNIST(DIR, train=False, transform=transforms.ToTensor())
+    valid_dataset = torch.utils.data.Subset(valid_dataset, indices=range(N_VALID_EXAMPLES))
+    valid_sampler = torch.utils.data.distributed.DistributedSampler(
+        dataset=valid_dataset, shuffle=False
+    )
+
+    train_loader = torch.utils.data.DataLoader(
+        train_dataset,
+        sampler=train_sampler,
+        batch_size=BATCHSIZE,
+        shuffle=False,
+    )
+    valid_loader = torch.utils.data.DataLoader(
+        valid_dataset,
+        sampler=valid_sampler,
+        batch_size=BATCHSIZE,
+        shuffle=False,
+    )
+
+    return train_loader, valid_loader, train_sampler, valid_sampler
+
+
+def objective(single_trial, device_id):
+    device = torch.device("cuda", device_id)
+    trial = optuna.integration.TorchDistributedTrial(single_trial, device=device)
+
+    # Generate the model.
+    model = DDP(define_model(trial).to(device), device_ids=[device])
+
+    # Generate the optimizers.
+    optimizer_name = trial.suggest_categorical("optimizer", ["Adam", "RMSprop", "SGD"])
+    lr = trial.suggest_float("lr", 1e-5, 1e-1, log=True)
+    optimizer = getattr(optim, optimizer_name)(model.parameters(), lr=lr)
+
+    # Get the FashionMNIST dataset.
+    train_loader, valid_loader, train_sampler, valid_sampler = get_mnist()
+
+    accuracy = 0
+    # Training of the model.
+    for epoch in range(EPOCHS):
+        model.train()
+        # Shuffle train dataset.
+        train_sampler.set_epoch(epoch)
```

Comment:
How about using the `sampler` attribute in the data loader?  
```suggestion
        train_sampler.sampler.set_epoch(epoch)
```

Then we can simplify `get_mnist` since we don't need `train_sampler, valid_sampler`.

Reply:
Do you mean `train_loader.sampler.set_epoch(epoch)` instead of `train_sampler.sampler.set_epoch(epoch)`? Yeah, indeed `train_sampler` is already passed to `train_loader` inside `get_mnist` so there's no need to return samplers from `get_mnist`. If that's what you mean, I can commit a change.

Reply:
Oh, sorry, Yes, you are right!